src/Chapter3.hs

```diff
@@ -480,6 +510,41 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Building = Church 
+              | Library 
+              deriving (Show)
+data Castle = MkCastle 
+            { castleName :: String,
+              casWalls :: Bool
+            } deriving (Show)
+data House = MkHouse {housePop :: Int} deriving (Show)
+
+data City = HasCastle Castle Building [House] 
+          | NoCastle Building [House]
+
+buildCastle :: City -> String -> City
+buildCastle (NoCastle b hs) castleName = HasCastle castle b hs 
+                                        where castle = MkCastle castleName False
+buildCastle (HasCastle _ b hs) castleName = HasCastle castle b hs 
+                                        where castle = MkCastle castleName False
+
+buildHouse :: City -> City
+buildHouse (NoCastle b hs) = (NoCastle b (h:hs))
+                          where h = MkHouse 0
+buildHouse (HasCastle c b hs) = (HasCastle c b (h:hs))
+                          where h = MkHouse 0
+
+census :: [House] -> Int
+census [] = 0
+census ((MkHouse pop):hs) = pop + census hs
+
+buildWalls :: City -> City
+buildWalls (HasCastle (MkCastle castleName _) b hs) | census hs >= 10 = HasCastle (MkCastle castleName True) b hs
+                                                    | otherwise = error "Not enough people to build walls!"
+buildWalls _ = error "You need a castle to build walls!"
```

Comment:
Interesting idea of using `error`. But, actually, it is not always desirable to produce runtime errors. If the city doesn't satisfy all checks, you have several options here:

1. Return `City` without changes.
2. Return `Maybe City` where `Nothing` means that wasn't possible to build walls.
3. Return `Either String City` with `String` being the reason why it wasn't possible to build walls (e.g. no castle or not enough people)